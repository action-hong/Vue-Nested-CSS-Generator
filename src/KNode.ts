export class KNode {
  tag: string
  clazz: string
  children: KNode[]
  constructor(tag: string, clazz: string) {
    this.tag = tag
    this.clazz = clazz
    this.children = []
  }

  addKNode(kNode: KNode) {
    this.children.push(kNode)
  }

  getCSS(depth = 0, includeTag = false): string {
    const inner = `${this.children.map(i => i.getCSS(includeTag || this.clazz ? depth + 1 : depth, includeTag)).join('\n')}`

    if (!includeTag && !this.clazz) {
      return inner
    }
    else {
      const space = ' '.repeat(2 * depth)
      const selector = `${space}${this.tag}${this.clazz}`
      return `${selector} {
  
  ${inner}
  ${space}}`
    }
  }

  isSame(kNode: KNode) {
    if (this.tag !== kNode.tag)
      return false
    if (this.clazz !== kNode.clazz)
      return false
    if (this.children.length !== kNode.children.length)
      return false
    for (let i = 0; i < this.children.length; i++) {
      if (!this.children[i].isSame(kNode.children[i]))
        return false
    }

    return true
  }
}
