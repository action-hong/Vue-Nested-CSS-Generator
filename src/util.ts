import type { SFCTemplateBlock } from '@vue/compiler-sfc'
import type { AttributeNode, DirectiveNode, TemplateNode } from '@vue/compiler-core'
import { compileTemplate } from '@vue/compiler-sfc'

import { KNode } from './KNode'

type ElementNode = SFCTemplateBlock['ast']

export interface GenerateCSSOption {
  includeTag: boolean
  openingTag: string
  closingTag: string
}

interface FileOption {
  source: string
  id: string
  filename: string
}

export function parseSFCTemplateBlock(node: SFCTemplateBlock['ast']) {
  const kNode = new KNode('', '', node.tag)
  walk(node, kNode)
  return kNode
}

export function isAttributeNode(node: AttributeNode | DirectiveNode): node is AttributeNode {
  return !!(node as AttributeNode).value
}

function walk(node: SFCTemplateBlock['ast'], parentKNode: KNode, depth = 0) {
  if (node.tag && (node.tagType === 0 || node.tagType === 1)) {
    const clazz = node.props.find(item => item.name === 'class')
    // 组件时，tag直接为设置为空
    const tag = node.tagType === 0 ? node.tag : ''
    let clz = ''
    if (clazz && isAttributeNode(clazz)) {
      const content = clazz.value?.content.trim()
      if (content)
        clz = `.${content.replace(/\s/g, '.')}`
    }
    const t = new KNode(tag === 'template' ? '' : tag, clz, node.tag)

    parentKNode.addKNode(t)
    parentKNode = t
  }
  if (node.children)
    node.children.forEach(n => walk(n as ElementNode, parentKNode, depth + 1))

  // 查看node.children下是否有重复的
  let i = parentKNode.children.length - 1
  while (i > 0) {
    const index = parentKNode.children.findIndex(item => item.isSame(parentKNode.children[i]))
    if (index !== -1 && index !== i) {
      // 有不同的
      parentKNode.children.splice(i, 1)
    }
    i--
  }
}

export function generateCSS(fileOption: FileOption, {
  closingTag = '</style>',
  openingTag = '<style lang="scss" scoped>',
  includeTag = false,
} = {}) {
  const template = transformSourceToTemplate(fileOption)

  if (template) {
    let css = parseSFCTemplateBlock(template as TemplateNode)
      .getCSS(0, includeTag)
    css = `${openingTag}
${css}
${closingTag}`

    return css
  }
}

function transformSourceToTemplate(fileOption: FileOption) {
  const result = compileTemplate({
    ...fileOption,
    compilerOptions: {
      hoistStatic: false,
    },
  })

  const template = result.ast?.children.find(item => (item as any).tag === 'template')

  return template
}
