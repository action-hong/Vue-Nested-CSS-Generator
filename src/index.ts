import { compileTemplate } from '@vue/compiler-sfc'
import type { TemplateNode } from '@vue/compiler-core'
import * as vscode from 'vscode'
import { parseSFCTemplateBlock } from './util'

const command = 'vue-generate-nested-css.generate'

function generateCSS() {
  const editor = vscode.window.activeTextEditor
  if (!editor)
    return

  const source = editor.document.getText()

  const result = compileTemplate({
    source,
    compilerOptions: {
      hoistStatic: false,
    },
    filename: editor.document.fileName,
    id: editor.document.fileName,
  })

  const template = result.ast?.children.find(item => (item as any).tag === 'template')

  if (template) {
    const css = parseSFCTemplateBlock(template as TemplateNode).getCSS()
    vscode.window.showInformationMessage('生成内容已复制到粘贴板')
    // 复制到
    vscode.env.clipboard.writeText(css)
  }
}

export function activate(ctx: vscode.ExtensionContext) {
  ctx.subscriptions.push(vscode.commands.registerCommand(
    command,
    () => {
      generateCSS()
    },
  ))
}

export function deactivate() {

}
