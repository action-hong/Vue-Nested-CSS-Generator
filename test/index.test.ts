import fs from 'fs'
import { describe, expect, it } from 'vitest'
import { generateCSS } from '../src/util'

function read(name: string) {
  return fs.readFileSync(`./test/template/${name}.vue`, 'utf-8')
}

function css(name: string, flag = false) {
  const source = read(name)
  const css = generateCSS({
    source,
    filename: name,
    id: name,
  }, {
    includeTag: flag,
  })
  return css || ''
}

describe('should', () => {
  // it('static', () => {
  //   expect(css('static')).toMatchSnapshot()
  //   expect(css('static', true)).toMatchSnapshot()
  // })

  // it('component', () => {
  //   expect(css('component')).toMatchSnapshot()
  //   expect(css('component', true)).toMatchSnapshot()
  // })

  it('complex', () => {
    expect(css('complex')).toMatchSnapshot()
    expect(css('complex', true)).toMatchSnapshot()
  })
})
