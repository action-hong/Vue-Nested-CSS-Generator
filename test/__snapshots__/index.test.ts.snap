// Vitest Snapshot v1

exports[`should > complex 1`] = `
"<style lang=\\"scss\\" scoped>
.container {

  .title {

  }
  .btn {

  }
}
</style>"
`;

exports[`should > complex 2`] = `
"<style lang=\\"scss\\" scoped>
div.container {

  header.title {

    ul {

    }
  }
  main {

    p {

      button.btn {

      }
    }
  }
}
</style>"
`;
